File: Byte_Compressor/src/core/preprocessor.js
/*
 * ByteCompressor - Preprocessing Transforms
 *
 * Pipeline: Input -> Delta -> BWT -> MTF -> RLE -> ANS
 *
 * Why this beats ZIP:
 *   ZIP uses LZ77 (sliding window dictionary) which misses long-range patterns.
 *   BWT groups similar contexts together, MTF converts them to small integers,
 *   and RLE collapses runs - creating an ideal low-entropy stream for ANS.
 */

/* ── Burrows-Wheeler Transform (BWT) ── */

function bwtForward(input) {
    const n = input.length;
    if (n === 0) return { output: Buffer.alloc(0), primaryIndex: 0 };

    // Double the input buffer to eliminate modulo in cyclic comparisons
    // This is the key optimization: comparison without expensive % operator
    const doubled = Buffer.alloc(n * 2);
    input.copy(doubled, 0);
    input.copy(doubled, n);

    const indices = new Int32Array(n);
    for (let i = 0; i < n; i++) indices[i] = i;

    // Sort cyclic rotations using doubled buffer (no modulo needed)
    indices.sort((a, b) => {
        for (let k = 0; k < n; k++) {
            const diff = doubled[a + k] - doubled[b + k];
            if (diff !== 0) return diff;
        }
        return 0;
    });

    // Extract last column and find primary index
    const output = Buffer.alloc(n);
    let primaryIndex = 0;
    for (let i = 0; i < n; i++) {
        if (indices[i] === 0) {
            primaryIndex = i;
            output[i] = input[n - 1];
        } else {
            output[i] = input[indices[i] - 1];
        }
    }
    return { output, primaryIndex };
}

function bwtInverse(input, primaryIndex) {
    const n = input.length;
    if (n === 0) return Buffer.alloc(0);

    // Count occurrences
    const counts = new Int32Array(256);
    for (let i = 0; i < n; i++) counts[input[i]]++;

    // Cumulative counts
    const cumul = new Int32Array(256);
    let sum = 0;
    for (let i = 0; i < 256; i++) {
        cumul[i] = sum;
        sum += counts[i];
    }

    // Build transform vector
    const T = new Int32Array(n);
    const running = Int32Array.from(cumul);
    for (let i = 0; i < n; i++) {
        T[running[input[i]]++] = i;
    }

    // Reconstruct original
    const output = Buffer.alloc(n);
    let idx = primaryIndex;
    for (let i = 0; i < n; i++) {
        idx = T[idx];
        output[i] = input[idx];
    }
    return output;
}

/* ── Move-to-Front Transform (MTF) ── */

function mtfEncode(input) {
    const list = Array.from({ length: 256 }, (_, i) => i);
    const output = Buffer.alloc(input.length);

    for (let i = 0; i < input.length; i++) {
        const ch = input[i];
        let rank = 0;
        while (list[rank] !== ch) rank++;
        output[i] = rank;
        // Move to front
        list.splice(rank, 1);
        list.unshift(ch);
    }
    return output;
}

function mtfDecode(input) {
    const list = Array.from({ length: 256 }, (_, i) => i);
    const output = Buffer.alloc(input.length);

    for (let i = 0; i < input.length; i++) {
        const rank = input[i];
        const ch = list[rank];
        output[i] = ch;
        list.splice(rank, 1);
        list.unshift(ch);
    }
    return output;
}

/* ── Delta Encoding ── */

function deltaEncode(input) {
    const output = Buffer.alloc(input.length);
    if (input.length === 0) return output;
    output[0] = input[0];
    for (let i = 1; i < input.length; i++) {
        output[i] = (input[i] - input[i - 1]) & 0xFF;
    }
    return output;
}

function deltaDecode(input) {
    const output = Buffer.alloc(input.length);
    if (input.length === 0) return output;
    output[0] = input[0];
    for (let i = 1; i < input.length; i++) {
        output[i] = (output[i - 1] + input[i]) & 0xFF;
    }
    return output;
}

/* ── Run-Length Encoding (RLE) ── */

function rleEncode(input) {
    const parts = [];
    let i = 0;
    while (i < input.length) {
        const current = input[i];
        let run = 1;
        while (i + run < input.length && input[i + run] === current && run < 256) {
            run++;
        }
        parts.push(current, run - 1);
        i += run;
    }
    return Buffer.from(parts);
}

function rleDecode(input) {
    const parts = [];
    for (let i = 0; i + 1 < input.length; i += 2) {
        const value = input[i];
        const count = input[i + 1] + 1;
        for (let j = 0; j < count; j++) parts.push(value);
    }
    return Buffer.from(parts);
}

module.exports = {
    bwtForward, bwtInverse,
    mtfEncode, mtfDecode,
    deltaEncode, deltaDecode,
    rleEncode, rleDecode
};

File: Byte_Compressor/src/core/preprocessor.c
/*
 * ByteCompressor - Preprocessing Transforms Implementation
 */

#include "preprocessor.h"
#include <string.h>
#include <stdlib.h>

/* ═══════════════════════════════════════════════════
 *  Burrows-Wheeler Transform (BWT)
 *  Groups bytes that share the same context together,
 *  dramatically improving compression on structured data.
 * ═══════════════════════════════════════════════════ */

/* Comparison context for qsort */
static const uint8_t *bwt_data;
static size_t bwt_size;

static int bwt_compare(const void *a, const void *b) {
    int32_t ia = *(const int32_t *)a;
    int32_t ib = *(const int32_t *)b;
    for (size_t k = 0; k < bwt_size; k++) {
        int ca = bwt_data[(ia + k) % bwt_size];
        int cb = bwt_data[(ib + k) % bwt_size];
        if (ca != cb) return ca - cb;
    }
    return 0;
}

int32_t bwt_forward(const uint8_t *input, uint8_t *output, size_t size, int32_t *workspace) {
    if (size == 0) return 0;

    /* Build suffix array via sorting */
    bwt_data = input;
    bwt_size = size;
    for (size_t i = 0; i < size; i++) {
        workspace[i] = (int32_t)i;
    }
    qsort(workspace, size, sizeof(int32_t), bwt_compare);

    /* Extract last column and find primary index */
    int32_t primary = 0;
    for (size_t i = 0; i < size; i++) {
        if (workspace[i] == 0) {
            primary = (int32_t)i;
            output[i] = input[size - 1];
        } else {
            output[i] = input[workspace[i] - 1];
        }
    }
    return primary;
}

void bwt_inverse(const uint8_t *input, uint8_t *output, size_t size, int32_t primary_idx) {
    if (size == 0) return;

    /* Count occurrences */
    int32_t counts[256] = {0};
    for (size_t i = 0; i < size; i++) {
        counts[input[i]]++;
    }

    /* Cumulative counts */
    int32_t sum = 0;
    int32_t cumul[256];
    for (int i = 0; i < 256; i++) {
        cumul[i] = sum;
        sum += counts[i];
    }

    /* Build transform vector */
    int32_t *T = (int32_t *)malloc(size * sizeof(int32_t));
    if (!T) return;

    int32_t running[256];
    memcpy(running, cumul, sizeof(cumul));
    for (size_t i = 0; i < size; i++) {
        T[running[input[i]]++] = (int32_t)i;
    }

    /* Reconstruct original */
    int32_t idx = primary_idx;
    for (size_t i = 0; i < size; i++) {
        idx = T[idx];
        output[i] = input[idx];
    }

    free(T);
}

/* ═══════════════════════════════════════════════════
 *  Move-to-Front Transform (MTF)
 *  Converts repeated nearby symbols into small integers (0,1,2...)
 *  which have very low entropy → excellent for ANS.
 * ═══════════════════════════════════════════════════ */

void mtf_encode(const uint8_t *input, uint8_t *output, size_t size) {
    uint8_t list[256];
    for (int i = 0; i < 256; i++) list[i] = (uint8_t)i;

    for (size_t i = 0; i < size; i++) {
        uint8_t ch = input[i];
        uint8_t rank = 0;

        /* Find position in list */
        while (list[rank] != ch) rank++;
        output[i] = rank;

        /* Move to front */
        memmove(list + 1, list, rank);
        list[0] = ch;
    }
}

void mtf_decode(const uint8_t *input, uint8_t *output, size_t size) {
    uint8_t list[256];
    for (int i = 0; i < 256; i++) list[i] = (uint8_t)i;

    for (size_t i = 0; i < size; i++) {
        uint8_t rank = input[i];
        uint8_t ch = list[rank];
        output[i] = ch;

        /* Move to front */
        memmove(list + 1, list, rank);
        list[0] = ch;
    }
}

/* ═══════════════════════════════════════════════════
 *  Delta Encoding
 *  Stores differences between consecutive values.
 *  Ideal for slowly-changing telemetry (temperature, voltage).
 * ═══════════════════════════════════════════════════ */

void delta_encode(const uint8_t *input, uint8_t *output, size_t size) {
    if (size == 0) return;
    output[0] = input[0];
    for (size_t i = 1; i < size; i++) {
        output[i] = input[i] - input[i - 1]; /* wraps modulo 256 */
    }
}

void delta_decode(const uint8_t *input, uint8_t *output, size_t size) {
    if (size == 0) return;
    output[0] = input[0];
    for (size_t i = 1; i < size; i++) {
        output[i] = output[i - 1] + input[i];
    }
}

/* ═══════════════════════════════════════════════════
 *  Run-Length Encoding (RLE)
 *  Collapses consecutive identical bytes.
 *  Format: [byte, count-1] for runs, or [byte, 0] for singles.
 *  Special: if count > 255, output multiple runs.
 * ═══════════════════════════════════════════════════ */

size_t rle_encode(const uint8_t *input, size_t size, uint8_t *output, size_t out_capacity) {
    size_t out_pos = 0;
    size_t i = 0;

    while (i < size) {
        uint8_t current = input[i];
        size_t run = 1;
        while (i + run < size && input[i + run] == current && run < 256) {
            run++;
        }

        if (out_pos + 2 > out_capacity) return 0; /* overflow */
        output[out_pos++] = current;
        output[out_pos++] = (uint8_t)(run - 1);
        i += run;
    }
    return out_pos;
}

size_t rle_decode(const uint8_t *input, size_t size, uint8_t *output, size_t out_capacity) {
    size_t out_pos = 0;
    size_t i = 0;

    while (i + 1 < size) {
        uint8_t value = input[i];
        uint8_t count = input[i + 1] + 1;
        i += 2;

        if (out_pos + count > out_capacity) return 0;
        memset(output + out_pos, value, count);
        out_pos += count;
    }
    return out_pos;
}
